feat: REST API article (#245)
shuttle-hq/www: _blog/2024-01-30-write-a-rest-api-rust.mdx

Interested in checking out what the final code should look like? You can find it [here.](https://github.com/joshua-mo-143/shuttle-axum-example/tree/main)

## Getting started

### Project initialisation

If you haven’t already, make sure you have Rust and Cargo installed! You can find install instructions [here.](https://www.rust-lang.org/tools/install)

First of all we’ll want to initialise our project with `cargo shuttle init` (required `cargo-shuttle` to be installed). For the project name we’ll be using `shuttle-example-axum`. 

Then we’ll want to set up our dependencies with the following shell snippet:

```rust
cargo add sqlx -F postgres,runtime-tokio-rustls
cargo add shuttle-shared-db -F sqlx,postgres
cargo add serde -F derive
```

### Adding a Database & Migrations in Rust

To add a database to our project, all we need to do is to add the annotation to our main function - which allows us to automatically grab a SQLx Postgres pool:

```rust
#[shuttle_runtime::main]
async fn main(
    #[shuttle_shared_db::Postgres] db: PgPool,
) -> shuttle_axum::ShuttleAxum {
    // .. your code here
}
```

> Interested in just getting the connection string? You can disable the `sqlx` feature of `shuttle-shared-db` and instead change the type to `String`, then connect to your PgPool. You can find more information about this 

Reviewer: Not sure if the last sentence on this row is complete. Where can we about more information about this?

> When writing our migrations, since we’re using only a single migrations file we can use `include_str!()` to include the whole file’s text content as a string then execute it with the database pool.

Reviewer: nit:
```suggestion
Since we’re using only a single migrations file for our migrations, we can use `include_str!()` to include the whole file’s text content as a string and then execute it with the database pool.
```

For the migrations, we can make a `migrations.sql` file in the project root:

```sql
-- migrations.sql

> CREATE TABLE IF NOT EXISTS Users (

Reviewer: nit: not sure if it is that common to start table names with capital letter. I think down the line the sql engine is not case sensitive, so it doesn't really matter.

```

Note that here, we will want to open a `Shuttle.toml` file in our project root and add the migrations file so that when we deploy, it will automatically include the migrations file:

```toml
assets = ["migrations.sql"]
```

We can add the database to our application by wrapping it in a struct and adding it to the application as shared state. The only requirement to add shared state is that the state must implement the `Clone` trait - which we can do here:

```rust
#[derive(Clone)]
pub struct AppState {
    db: PgPool,
}
```

Note that if you’re using a struct that is unable to use Clone (for example because of generic trait bounds), you can wrap the type in a `std::sync::Arc`.

Then we need to add it to our API:

```rust
let state = AppState { db };

let router = Router::new().route("/", get(hello_world)).with_state(state);
```

### Writing routes

When writing your routes, Axum will accept anything that implements the `IntoResponse` trait (which means the type can be turned into a HTTP response). Many primitives already have `IntoResponse` implemented so you can return things like `i32` or a `String` (or a `'static &str`) without being required to implement the trait. 

> By using Axum’s `extractors`, we can extract information from the HTTP request and declare them as handler parameters: for example, adding a `State` extractor means we want to include the state that we added to our `axum::Router`. We can g

Reviewer: It seems like the last sentence is incomplete.

    let res = match sqlx::query_as::<_, User>("SELECT * FROM USERS").fetch_all(&state.db).await {
        Ok(res) => res,
        Err(e) => {
            return Err((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()));
        }
    };

    Ok(Json(res))
}
```

After this, now that we’ve written our initial route for fetching all records, we can also write our other routes. For retrieving record by ID, we can use the `Path` extractor which extracts a dynamic URL slug (this will be shown later on):

```rust
async fn retrieve_record_by_id(
    State(state): State<AppState>,
    Path(id): Path<i32>
) -> Result<impl IntoResponse, impl IntoResponse> {
    let res = match
        sqlx
            ::query_as::<_, User>("SELECT * FROM USERS WHERE id = $1")
            .bind(id)
            .fetch_one(&state.db).await
    {
        Ok(res) => res,
        Err(e) => {
            return Err((StatusCode::INTERNAL_SERVER_ERROR, e.to_string()));
        }
    };

    Ok(Json(res))
}
```

Note that for the SQL query, we also bind a parameter to the query itself. By using `bind` variables exclusively when we need to add data from the JSON body, SQL injection can be easily prevented!

> Likewise below, we can also create a `UserSubmission` struct to represent a JSON request body for a user submission. We need to implement `serde::Deserialize` to be able to successfully extract it from the request body. We can also can add a handler function for deleting records by ID in this snippet below:

Reviewer: ```suggestion
Likewise, below, we can also create a `UserSubmission` struct to represent a JSON request body for a user submission. We need to implement `serde::Deserialize` to be able to successfully extract it from the request body. We can also add a handler function for deleting records by ID in this snippet below:
```
